File: articles/alan-turing-computation-and-the-machine-model-of-reasoning/java/UnaryIncrement.java
public class UnaryIncrement {
    public static String unaryIncrement(String input) {
        StringBuilder tape = new StringBuilder(input);
        int i = 0;
        while (i < tape.length() && tape.charAt(i) == '1') {
            i++;
        }
        if (i == tape.length()) {
            tape.append('_');
        }
        tape.setCharAt(i, '1');
        if (i + 1 == tape.length()) {
            tape.append('_');
        }
        return tape.toString();
    }

    public static void main(String[] args) {
        System.out.println("incremented_tape=" + unaryIncrement("111_"));
    }
}

File: articles/alan-turing-computation-and-the-machine-model-of-reasoning/cpp/unary_increment.cpp
#include <iostream>
#include <string>

std::string unary_increment(std::string tape) {
    std::size_t i = 0;
    while (i < tape.size() && tape[i] == '1') {
        i++;
    }
    if (i == tape.size()) {
        tape.push_back('_');
    }
    tape[i] = '1';
    if (i + 1 == tape.size()) {
        tape.push_back('_');
    }
    return tape;
}

int main() {
    std::cout << "incremented_tape=" << unary_increment("111_") << "\n";
    return 0;
}
